libs/cardano-ledger-core/src/Cardano/Ledger/DRepDistr.hs

```diff
@@ -0,0 +1,222 @@
+  DRepPulser ::
+    (ans ~ Map (DRep c) (CompactForm Coin), m ~ Identity) =>
+    !Int ->
+    !(UMap c) ->
+    !(Set (Credential 'Voting c)) ->
+    -- purposely not strict
+    (VMap VB VP (Credential 'Staking c) (CompactForm Coin)) ->
+    !ans ->
+    DRepPulser c m ans
+
+deriving instance Show ans => Show (DRepPulser c m ans)
+
+instance Typeable c => NoThunks (DRepPulser c Identity (Map (DRep c) (CompactForm Coin))) where
+  showTypeOf _ = "DRepPulser"
+  wNoThunks ctxt (DRepPulser n um _drep balance ans) =
+    allNoThunks
+      [ noThunks ctxt n
+      , noThunks ctxt um
+      , --   , noThunks ctxt drep -- Here there be thunks
+        noThunks ctxt balance
+      , noThunks ctxt ans
+      ]
+
+instance NFData (DRepPulser c Identity (Map (DRep c) (CompactForm Coin))) where
+  rnf (DRepPulser n1 um drep bal ans) = seq (rnf n1) (seq (rnf um) (seq (rnf drep) (seq (rnf bal) (rnf ans))))
+
+instance Pulsable (DRepPulser c) where
+  done (DRepPulser _ _ _ m _) = VMap.null m
+  current (DRepPulser _ _ _ _ ans) = ans
+  pulseM (ll@(DRepPulser _ _ _ balance _)) | VMap.null balance = pure ll
+  pulseM (DRepPulser n um regDreps balance ans) =
+    let !(!steps, !balance') = VMap.splitAt n balance
+        ans' = foldlWithKey (accumDRepDistr um regDreps) ans steps
+     in Identity (DRepPulser n um regDreps balance' ans')
+
+-- ===================================================================================
+-- The data structure we store in the Ledger State. In some Rule under the TICK rule
+-- we will update this data structure each time we Tick. Hopefully at some point
+-- further Ticking becomes idempotent, and the DRep stake distribution is the result.
+
+-- | DRepDistr is what we store in the State, It is pulsed in some sub-Rule run by the TICK rule.
+--   We will need lots of instances (Generic,NoThunks,NFData,Eq,EncCBOR,DecCBOR) as
+--   as this type will be part of the Ledger State, and must be serialized, printed, subject to equality tests etc.
+data DRepDistr c
+  = DRPulsing !(DRepPulser c Identity (Map (DRep c) (CompactForm Coin)))
+  | DRComplete !(Map (DRep c) (CompactForm Coin))
+  deriving (Generic, NoThunks, NFData)
+
+instance Eq (DRepDistr c) where
+  DRComplete x == DRComplete y = x == y
+  DRComplete x == DRPulsing y = x == complete y
+  DRPulsing x == DRComplete y = complete x == y
+  DRPulsing x == DRPulsing y = complete x == complete y
+
+instance Show (DRepDistr c) where
+  show (DRComplete m) = "(DRComplete " ++ show m ++ ")"
+  show (DRPulsing p) = "(DRPulsing " ++ show p ++ ")"
+
+instance Crypto c => EncCBOR (DRepDistr c) where
+  encCBOR (DRComplete x) = encode (Rec DRComplete !> To x)
+  encCBOR (DRPulsing x) = encode (Rec DRComplete !> To (complete x))
+
+instance Crypto c => DecCBOR (DRepDistr c) where
+  decCBOR = decode (RecD DRComplete <! From)
+
+instance ToExpr (DRepDistr era) where
+  toExpr (DRComplete x) = App "DRComplete" [toExpr x]
+  toExpr (DRPulsing x) = App "DRComplete" [toExpr (complete x)]
```

Comment:
Should this be App "DRPulsing" ?

Reply:
Actually it should be "DRComplete".  Notice how it completes the pulser when turning it into an Expr.  This is a pattern we use a lot. Not that in (DRPulsing x) 'x' has a very complicated internal structure, with internal state, and what we reay want is to compare two DRepDistr objects and report the subparts that are different.  That wouldn't really work, because two pulsers that have puled a different number of times, would have different internal structures, but the same final result. So we run every (DRPulsing x) to completion and compare the final results.

Reply:
This make sense, thank you for the explanation